src/power_grid_model/core/dataset_definitions.py

```diff
@@ -72,32 +72,33 @@ class ComponentType(str, Enum, metaclass=_MetaEnum):
     asym_power_sensor = "asym_power_sensor"
     fault = "fault"
 
+# fmt: on
 
 # pylint: enable=invalid-name
 
+ComponentTypeLike: TypeAlias = ComponentType | str
+ComponentTypeVar = TypeVar("ComponentTypeVar", bound=ComponentTypeLike)  # helper used for type deduction
+
 
-def _str_to_datatype(data_type: Any) -> DatasetType:
+def _str_to_datatype(data_type: DatasetType | str) -> DatasetType:
     """Helper function to transform data_type str to DatasetType."""
-    if isinstance(data_type, str):
-        return DatasetType[data_type]
-    return data_type
+    if isinstance(data_type, DatasetType):
+        return data_type
+    return DatasetType[data_type]
 
 
-def _map_to_datatypes(data: Mapping[Any, Any]) -> dict[DatasetType, Any]:
+def _map_to_datatypes(data: Mapping[DatasetType | str, Any]) -> dict[DatasetType, Any]:
     """Helper function to map datatype str keys to DatasetType."""
     return {_str_to_datatype(key): value for key, value in data.items()}
 
 
-def _str_to_component_type(component: Any) -> ComponentType:
+def _str_to_component_type(component: ComponentTypeLike) -> ComponentType:
     """Helper function to transform component str to ComponentType."""
-    if isinstance(component, str):
-        return ComponentType[component]
-    return component
+    if isinstance(component, ComponentType):
+        return component
+    return ComponentType[component]
 
 
-def _map_to_component_types(data: Mapping[Any, Any]) -> dict[ComponentType, Any]:
+def _map_to_component_types(data: Mapping[ComponentTypeVar, Any]) -> dict[ComponentType, Any]:
```

Comment:
be as flexible as you need to be on the input side, and as restrictive as you can be on the output side

Reply:
NOTE: since this only affects private core functionality (private because it's prefixed `_`; core because it's not exposed in the `__init__` or other public modules outside the `core` module), this does not affect the user